### PR TITLE
feat: add ability to exclude branches from filtering and select days to filter by discovery type

### PR DESCRIPTION
**Pull request by ryancurrah**
This pull request adds the ability to select the days to filter by discovery type (branch, pr, tag). It also allows us to exclude branch names from being filtered like main or hotfix-* branches, since people would most likely want to keep old builds on those branches around for audit purposes.

### Testing done

- Wrote unit tests. 
- Run change in production with 10,000+ jobs configured to use it for over a year.
 
I was going to make this PR when I first modified the project almost two years ago but did not find the time.

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by ryancurrah** [comment]
> Is it possible to grab a plugin from this PR for beta testing? I've been waiting for this feature a quite a while and am excited to test it out.

@wiiNinja sorry for the delay. Here are the jars built with `openjdk 17.0.10` https://github.com/ryancurrah/scm-filter-aged-refs-plugin/releases/tag/v0.0.0-alpha

**Comment by TobiX** [comment]
Sorry, I somehow had notifications disabled for this repository and didn't see this PR. Trying to get around to reviewing it in the coming days/weeks...

File: bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java [hunk 1]
```diff
@@ -19,26 +19,27 @@
...
+     * @param branchRetentionDays retention period in days for branches
+     * @param prRetentionDays     retention period in days for pull requests
+     * @param tagRetentionDays    retention period in days for tags
+     * @param branchExcludeFilter space-separated list of branch name patterns to
+     *                            ignore. For example: release main hotfix-*
      */
     @DataBoundConstructor
-    public BitbucketAgedRefsTrait(String retentionDays) {
-        super(retentionDays);
+    public BitbucketAgedRefsTrait(
```
Comment (TobiX) [inline bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: Please don't add new oprtions to the `@DataBoundConstructor`, use `@DataBoundSetter`s instead.

File: bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java [hunk 2]
```diff
@@ -1,30 +1,58 @@
 package org.jenkinsci.plugins.scm_filter;
 
-import static org.assertj.core.api.Assertions.assertThat;
```
Comment (TobiX) [inline bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java:3]: Please don't switch the assertion framework.

File: bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java [hunk 3]
```diff
@@ -1,30 +1,58 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
 
 import com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource;
-import java.io.IOException;
-import java.io.InputStream;
 import jenkins.model.Jenkins;
-import org.junit.jupiter.api.Test;
```
Comment (TobiX) [inline bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java:9]: Please don't revert from JUnit Jupiter to JUnit 4.

**Comment by wiiNinja** [comment]
@ryancurrah @TobiX - Really appreciate you both. This is a feature that is much needed by our team. I'm going to try out the beta while waiting for the merge.

**Comment by ryancurrah** [comment]
We've been running it for over a year on enterprise Jenkins with no issues. I expect no issues. If you're using Bitbucket and trying to limit tags by age just know that it will use the commits age not the tags age. This is due to the Bitbucket API not having an endpoint to get a tags create date.

File: bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java (hunk 1 above)
Comment (ryancurrah) [reply bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: I'm not a Jenkins plugin expert, I am assuming you don't want to break the deserialization of existing `config.xml` files that don't have a value for the new parameters? The problem is there is no more `retentionDays` variable. So what should we do?

What if I changed it to...

```
    public BitbucketAgedRefsTrait() {
        super();
    }

    @DataBoundSetter
    public void setBranchRetentionDays(String branchRetentionDays) {
        super.setBranchRetentionDays(branchRetentionDays);
    }

    @DataBoundSetter
    public void setPrRetentionDays(String prRetentionDays) {
        super.setPrRetentionDays(prRetentionDays);
    }

    @DataBoundSetter
    public void setTagRetentionDays(String tagRetentionDays) {
        super.setTagRetentionDays(tagRetentionDays);
    }

    @DataBoundSetter
    public void setBranchExcludeFilter(String branchExcludeFilter) {
        super.setBranchExcludeFilter(branchExcludeFilter);
    }
```

File: bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java (hunk 2 above)
Comment (ryancurrah) [reply bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java]: Resolved.

File: bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java (hunk 3 above)
Comment (ryancurrah) [reply bitbucket-scm-filter-aged-refs/src/test/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTraitTest.java]: Resolved.

**Comment by ryancurrah** [comment]
Apologies for the delay. I resolved 2/3 issues. Not sure what to do about `@DataBoundConstructor` vs`@DataBoundSetter` I am hoping for a little more guidance.

File: bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java (hunk 1 above)
Comment (TobiX) [reply bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: If you removed a variable, it needs to stay in the class to not break the serialization format, in the base case with migration code to new variables... (Or keep the coarse-grained setting and hide the individual settings behind an "advanced" pop-out?)

Comment (TobiX) [reply bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: See https://www.jenkins.io/doc/developer/persistence/backward-compatibility/

Comment (rpaasche) [reply bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: As far as I see you only have to add something like:
```
protected transient Long retentionDays;

protected Object readResolve() {
  if (retentionDays!= null) {
    super.setBranchRetentionDays(retentionDays); 
    super.setPrRetentionDays(retentionDays);
    super.setTagRetentionDays(retentionDays);
    super.setBranchExcludeFilter(retentionDays);
  }
  return this;
}
```

Comment (ryancurrah) [reply bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: If I follow ☝️  do I revert my changes to `@DataBoundConstructor`?

Comment (ryancurrah) [reply bitbucket-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/BitbucketAgedRefsTrait.java:34]: Ok well I applied the changes based on feedback and the doc linked.

**Comment by ryancurrah** [comment]
@TobiX check this commit out, I think this may be what you were looking for https://github.com/jenkinsci/scm-filter-aged-refs-plugin/pull/26/commits/69cf0742b0273b7d75ee0df8ec6eecd07a9a322f

**Comment by TobiX** [comment]
After careful consideration, I will not merge this in its current state, sorry. I quite like the simplicity of the current trait and this would move other users who are happy with the current functionality to a more "complicated" configuration.

I'm open to adding more configurable traits, though. Since traits are quite easy to implement (for the developer) and add (for the user), I'd suggest to create new traits for branches, PRs and tags, respectively. I'd argue the branch filter might be useful enough to extend the current trait with it, though (might even implement it myself in the future). This would also neatly solve the "problem" with backward compatibility (since there would be none)

**Pull request closed without merging**